### PR TITLE
1st draft at HowTo

### PR DESCRIPTION
**Pull request by fieker**
please comment - or better improve

I tried to be harmless

### CODE REVIEW

**Comment by fieker** [comment]
Meant to be a draft only...

File: WhatIs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  In algebra, this is either not true or terribly inefficient (or impossible)
+  Take
+    Z/nZ integers modulo n, and matrices over it
+   
+   Then either
+    n is part of the type -> every function is recompiled for every n - which kills
+    all modular (Chenese remainder theorem  (CRT) based) algorithms
+
+    n is not part of the type, then it need to be elsewhere, ie. in the parent,
```
Comment (mohamed-barakat) [inline WhatIs]:
```suggestion
    n is not part of the type, then it needs to be elsewhere, i.e. in the parent,
```

File: HowTo.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
+# Oscar - how do we envision development
```
Comment (fingolfin) [inline HowTo.md]: IMHO the file should be renamed, perhaps to "PLAN.md" or "DEVELOPMENT.md" or so

File: HowTo.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ - number fields
+ - (abelian) groups
+ - ... and MUCH more
+As well as specialised stuff
+ - polymake
+ - Singular
+ - Gap
+
+So: please use it. It is safe to assume all can be improved, however,
+then we won't get anywhere. For preference: interoperability is much more
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
So: please use it. It is safe to assume all can be improved, however, if we try to perfect every
single line of code again and again, we we won't get anywhere; there is a balance to be found.
For preference: interoperability is much more
```

File: HowTo.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+important than speed, readability is much more important than speed. Having
+said that: of course, sometimes pure speed matters, but not nearly as often as
+people think.
+
+The infrastructure is incomplete, e.g. we do not have
+ - varieties, (we're started with curves recently)
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
 - varieties, (we've started with curves recently)
```

File: HowTo.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+At this point we have, mathematically speaking, almost nothing. This has to
+change, quickly.  We work in a team of ~ 10(?) people.
+We lack infrastructure as everyone will notice immediately.
+As a result, the classical
+   I work for 6 month in a separate repo on a branch and then will dazzle you with perfect
+   code and cool examples
+is not going to work for now. It will result in everyone fixing the same
+infrastructure problems over and over again. Please consider to work, e.g. in
+a file/directory  in Oscar/examples and push on a regular basis, even, or in
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
a file/directory  in `Oscar/examples` and push on a regular basis, even, or in
```

File: HowTo.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
+# Oscar - how do we envision development
+
+## What Do We Have:
+
+We have a large code-base for infrastructure in place, comprising at least
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
We have a large codebase for infrastructure in place, comprising at least
```

File: HowTo.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+a file/directory  in Oscar/examples and push on a regular basis, even, or in
+particular, incomplete code. Break it down into small pull requests.
+ - it is in examples, so no-one is accidentally going to automatically use it
+ - it is in the main repo, so everyone can see it and try to use it, producing feedback
+ - infrastructure, in Oscar at least, can be added in the same repo, immediately and thus
+   is available to others, immediately. (feedback, ...)
+ - Talk to us - as early as possible. There is a chance that what you're trying to
+   do is already there - in a different package maybe, not linked in, ....
+ - Talk to us! We are
+   - on slack: send an email to webmaster-oscar@mathematik.uni-kl.de
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
   - on slack: send an email to webmaster-oscar@mathematik.uni-kl.de to be added
```

File: HowTo.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+infrastructure problems over and over again. Please consider to work, e.g. in
+a file/directory  in Oscar/examples and push on a regular basis, even, or in
+particular, incomplete code. Break it down into small pull requests.
+ - it is in examples, so no-one is accidentally going to automatically use it
+ - it is in the main repo, so everyone can see it and try to use it, producing feedback
+ - infrastructure, in Oscar at least, can be added in the same repo, immediately and thus
+   is available to others, immediately. (feedback, ...)
+ - Talk to us - as early as possible. There is a chance that what you're trying to
+   do is already there - in a different package maybe, not linked in, ....
+ - Talk to us! We are
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
 - Talk to us!
```

File: HowTo.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+   - on slack: send an email to webmaster-oscar@mathematik.uni-kl.de
+   - on oscar-dev@mathematik.uni-kl.de
+
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
   - whenever you meet us (at the uni, during conferences, ...)
```

File: HowTo.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+   do is already there - in a different package maybe, not linked in, ....
+ - Talk to us! We are
+   - on slack: send an email to webmaster-oscar@mathematik.uni-kl.de
+   - on oscar-dev@mathematik.uni-kl.de
+   
+
+# Practical Development
+
+For practical development:
+ - use ```Revise```
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
 - use [the `Revise` package](https://github.com/timholy/Revise.jl)
```

File: HowTo.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+ - start a file/ directory in Oscar/examples, call it Blah.jl
+
+Then ```Oscar.example("Blah.jl")``` will include it, so you don't have to worry about a path
+```Oscar.revise("Blah.jl")``` will put it under revise, ie. re-load automatically.
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
Then `Oscar.example("Blah.jl")` will include it, so you don't have to worry about a path;
and `Oscar.revise("Blah.jl")` will put it under revise, i.e. re-load it automatically whenever you make a change.
```

File: HowTo.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+ end
+
+ using .BlahModule
+``` 
+
+then you can even redefine struct's without quitting Julia (and restarting)
+
+
+Julia's convention is to use lower_case_and_underscores, not CamelCase for functions,
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
Julia's convention is to use `snake_case`, not `CamelCase` for function names,
```

File: HowTo.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+so that is the accepted style in Oscar.
+types (and their constructor) tend to be CamelCase. However, for me, as a user,
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
Types (and their constructor) tend to be in `CamelCase`. However, for me, as a user,
```

File: HowTo.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+In Oscar all objects have a default mathematical meaning and we have multiple dispatch. As
+a result
+ - IsSurjectiveMapForAbelianGroups
+ - IsSurjectiveMapForRing
+is not necessary
+ - issurjective
+will do, the different calling signatures will add the type info. If you have a ring map
+and want to test it as a map between additive/ multiplicative groups, then you'll have to 
+create a container for it... (we can always discuss exceptions)
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
first convert/coerce/wrap it (in)to a map between groups... (we can always discuss exceptions)
```

File: HowTo.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+We will have to add "forgetful functor" containers.
```
Comment (fingolfin) [inline HowTo.md]:
```suggestion
We will have to add "forgetful functor" wrappers.
```

File: README.md [hunk 16]
```diff
@@ -38,10 +38,10 @@ julia> using Oscar
  -----    -----    -----   -     -  -     -  
 
 ...combining (and extending) GAP, Hecke, Nemo, Polymake and Singular
-Version 0.4.0 ... 
+Version 0.5.1-DEV ... 
  ... which comes with absolutely no warranty whatsoever
 Type: '?Oscar' for more information
-(c) 2019-2020 by The Oscar Development Team
+(c) 2019-2021 by The Oscar Development Team
```
Comment (fingolfin) [inline README.md]: You have somehow collected a bunch of changes that shouldn't be in this PR. Can easily be fixed by rebasing it on `master` (but it needs to be the *current* master; i.e.: `git fetch origin && git rebase origin/master` assuming `origin` == this repository)

File: WhatIs.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
+# Oscar - what is the idea
+
+Oscar should be competing with (and beating) Magma and Sage in our areas of
```
Comment (fingolfin) [inline WhatIs.md]:
```suggestion
The ultimate goal for Oscar should be to compete with (and ideally beat) Magma and Sage in our areas of
```

File: WhatIs.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+expertise.
+
+Oscar is the innovative, next generation Computer Algebra System
+
+Oscar is written in Julia, but is not Julia, nor can it be. 
+ Example:
```
Comment (fingolfin) [inline WhatIs.md]:
```suggestion
 Some examples to illustrate what that means:
```

File: WhatIs.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+  Julias matrices are (c)-Arrays (of arbitrary dimension), parametrized by
```
Comment (fingolfin) [inline WhatIs.md]:
```suggestion
  Julias matrices are (C) arrays (of arbitrary dimension), parametrized by
```

**Review by fingolfin — COMMENTED** [review]
Some comments. You can directly apply them (individually or all at once).

**Comment by fieker** [comment]
how can I accept them in bulk? (apart from the re-base-ing)

**Pull request merged**